### DISCUSSION TITLE
Internship_Report_Vaishnavi

### DISCUSSION THREAD

**Discussion by Vaish2611** [petermr/petermr#25] [Show and tell]
**This channel conatins a brief report of my work that I did during my internship**

**Comment by Vaish2611** [comment]
Hi, I am Vaishnavi Jha, a 3rd-year undergrad doing a B.Tech in Biotechnology from Delhi Technological University. I am an research enthusiast, and hope to frame my path around computational biology.

This discussion channel is my medium to talk about all the subprojects I worked on during this 3-month internship under the guidance of Dr. Peter Murray Rust.

**Reply by EmanuelFaria** [reply]
Hi Vaishnavi, I finally got a chance to review your draft report.

I find it cumbersome to do that within this discussion page so I've made a [google doc where you can see my comments, here](https://docs.google.com/document/d/1O7184Ma4ZzIFsI8E1SsL9szXZazfrpqm/edit?usp=sharing&ouid=110062979113982375503&rtpof=true&sd=true).

Keep in mind, my background is in writing for business and marketing, not science so use my comments as tips only, and use your best judgment.

I hope this helps. :)

**Comment by Vaish2611** [comment]
## **PROJECT 1: Building tools for analysis of Terpene Classification Task:**
With the help of constant cmmunal disussion we are tryingto bild a machine learning tool that could identify terpenes from the images (source alex pico's wikipathway). 

## **Our Purpose:**
We needed to come to a confirmation that whether the extracted images on [wikipathway page](https://gladstone-bioinformatics.shinyapps.io/shiny-terpeneplant/) from various papers, articles, journels etc. consistis of Terpene Pathway.

For this purpose we asked various people with any background to annotate the CSV file so that we could analyze the database and get the needed confirmation.

## **Methodology:**

With the collected raw data following the use of statistical analysis based on inter-annotator agreement and correlation coefficient we analyzed the annotation and came to a certain conclusion.

## **Process:**

To formlate a basic structure for construction of the tool, we started with creating a simple rule book that an annotator could go through and fill out the coloumns accordingly: 

Intially the tabel to be filled looked like:

PMCID | Fig No. | Pathway | Metabolic | Pre-Terpene | Terpene | Terpene Synthase 
-- | -- | -- | -- | -- | -- | --
  |   |   |   |   |  |

With the help of constant communal discussion we came with the draft of [rules](https://github.com/petermr/CEVOpen/wiki/Rules-for-Classification-of-terpene-pathway) which we circulated across to get the inputs.

<img width="400" alt="image" src="https://user-images.githubusercontent.com/78601110/161517639-a06df958-407a-428c-8848-671f3db53a30.png">

To sucessfully operate on terpene classification task we took dataset from a trusted source. In Alex's website there are more than 300 pathway images some belong to terpene and others not. Therefore it suited our purpose the best, as we wanted to understnd the statistics behind terpene classification task and learn how difficult is gets for human to classifiy images to narrower domains like [pre- terpene, terpene] from broader domain like [Pathway, Metabolic].
The table below will give you a glimse of our data looks like for the experimentation.

![image](https://user-images.githubusercontent.com/78601110/161518545-c686b461-12d3-4699-9c7e-739a7ba37170.png)

The motive of our doing this was to find out how much difficult it is for humans to annotate or differentiate whether particular image is of a terpene or not or is it a pathway or not. We had circulated 30 images and got 8 annotations which was collected via github and everything was done on open source. A simple annotator file looked like this:

<img width="348" alt="image" src="https://user-images.githubusercontent.com/78601110/161519273-32ecf51e-1393-4983-bec4-593f4cbb74b3.png">

The annotated data from the 8 snnotators is then used for the statistical annalysis. You can find discription about the code in [here](https://github.com/petermr/CEVOpen/blob/master/statistical_ananlysis_2022/statistical_analysis_annotators.ipynb). 

## **Observation:**

- The annotations analysis gave us some exciting results, from which it was clear that people could differentiate between the basic terminologies. The only difficulty they faces was when the terminologies got specific like maximum disagreement was in case of terpene 
- The result also suggested the modifications for the rules i.e. terpene rules needed framing and sorting so that one could better analyze the image.
- The results also showed that images had maximum non-terpene and pre-terpene pathways. 

## **Future Aspects:** 

1. Rules need modification.
2. Another round of annotations with specific tabel should be circulated.

**Reply by EmanuelFaria** [reply]
PROJECT 1: Building tools for analysis of Terpene Classification Task:
> TITLE: If I understand your intention, I “think” your title here should more specifically describe what you’ve writing in the next section you’ve called “Purpose” (which you may want to change to “Objective”). If I’m not mistaken, general “Terpene Classification” may better describe the main project, of which the sub-project is Increment in its accomplishment (you’re working on a “tool” in the overall “toolkit” called docanalysis, I think). Maybe something like “Developing Tools for Machines to Classify Terpenes”?
> BACKGROUND: If I’ve understood the assignment so far, vetween “Project 1” and “Purpose”, consider adding a heading called “BACKGROUND”. Here you could start by explaining that what you’re writing about in this report, and how this sub-project you’ve participated in fits into the overall toolkit project as briefly as you can, and then link to the “Toolkit’s” description page, (which I think you and Roopa and Ananya will work on together and post separately).
> PURPOSE/OBJECTIVE: Here you could briefly explain that the project being described below is a key part of making a functional tool that will be part of a toolkit that will extract data from images… etc.. then go right to describing this project in the next section.
With the help of constant cmmunal disussion we are tryingto bild a machine learning toolkit that could identify terpenes from the images (source alex pico's wikipathway).
> Suggestions: 1) Link “wikipathway.org” to URL with best, shortest description and include “.org” in sentence. 2) Replace “could” with “will”. 3) Consider moving the part about “with the help of constant communal discussion“ to your “About Me” section and introduce yourself, the team and how you all worked remotely there so you can go right to the project here and in the (sub)sections below.
Our Purpose:
> Perhaps, “Objective” rather than Purpose?
We needed to come to a confirmation that whether the extracted images on wikipathway page from various papers, articles, journels etc. consistis of Terpene Pathway.
> This is good “We want..” statement.. consider adding a “So that… (outcome). for the purpose of… (being used by scientists to x without y?)
For this purpose we asked various people with any background to annotate the CSV file so that we could analyze the database and get the needed confirmation.
> I think this would be a great first sentence under “Methodology”. It might be useful to think about structuring each step something like: Do/did [x] so that we accomplish [y] (without [z] happening). Then you could replace “For this purpose” with “First,” and then list the steps.
Methodology:
With the collected raw data following the use of statistical analysis based on inter-annotator agreement and correlation coefficient we analyzed the annotation and came to a certain conclusion.
> Break this into at least two sentences. My layman brain got confused. ;)
Process:
To formlate a basic structure for construction of the tool, we started with creating a simple rule book that an annotator could go through and fill out the coloumns accordingly:
> This might be easier to explain as steps. Also I’m not sure if “construction” should be “development”, but Peter would know.
Intially the tabel to be filled looked like:

With the help of constant communal discussion we came with the draft of rules which we circulated across to get the inputs.
> Suggestion: Replace “we came with the draft of” with “our team drafted the following”. I don’t know what “circulated across to get the inputs” means.

To sucessfully operate on terpene classification task we took dataset from a trusted source. 
> … from a trusted source: wikipathways.org”
> This sentence feels to me like the beginning of a change of topic. If so, suggest you add a sub-sub-heading that defines to the reader what it is they’re about to read. 
> Here's an old tip for writing marketing copy that can apply in the different sections of this report and practically anything else you’ll write:
> 1. Tell them what you’re going to tell them.
> 2. Tell them.
> 3. Tell them what you told them.
> Another think I’ve learned the hard way is to always begin a paragraphs and sentences with the “WHY” before you get to the “WHAT” or “HOW”. Starting with the WHY gives the reader the context required to understand the significance of the WHAT and HOW. 
> You’ve sort of donE that here with “To successfully operate on terpene…”,, but I go into read the paragraph wondering WHY it’s important. So, maybe a stronger WHY could be something along the lines of, “It is important that our terpene classifications were x and not y, so we [did] “WHAT” by “HOW”.
> Here’s an example from my previous working in marketing real estate investments… there were about 4,000 investors who would receive a letter from me explaining an upcoming change to their investment. If I wrote it with the “WHY” last, like this: 
> “We are stopping your monthly interest payments so the money can be used to improve the properties and thereby increase the value of your investment” I’d get frantic emails and phone calls from about 15% of the audience who freaked out and stopped reading after “We are stopping your monthly interest payments”. Instead, I give them the context-creating WHY at the front of the sentence: “In order to increase the value of the properties, and thereby your investment, we are temporarily pausing your monthly payments.” I hope this example provides the “WHY” for you to remember this tip in all your writing. :)
In Alex's website there are more than 300 pathway images some belong to terpene and others not. Therefore it suited our purpose the best, as we wanted to understnd the statistics behind terpene classification task and learn how difficult is gets for human to classifiy images to narrower domains like [pre- terpene, terpene] from broader domain like [Pathway, Metabolic].
> “as we wanted to understnd the statistics behind terpene classification task and learn how difficult is gets for human to classifiy images” is the WHY in this sentence. I recommend starting with that.
 The table below will give you a glimse of our data looks like for the experimentation.

To find out just how difficult it is for humans to annotate or differentiate whether particular image is of a terpene is a pathway or not, we had circulated 30 images and got 8 annotations collected via GitHub. (Everything was done on open source basis.) 
A simple annotator file looked like this:

The annotated data from the 8 snnotators is then used for the statistical annalysis. You can find discription about the code in here.
Observations:
The annotations analysis gave us some exciting results, from which it was clear that people could differentiate between the basic terminologies. The only difficulty they faced was when the terminologies got specific, like maximum disagreement was in case of terpene
> I don’t know what “maximum disagreement” is a common statistical term that your audience would know, but if so that’s fine.
The result also suggested the modifications for the rules i.e. terpene rules needed framing and sorting so that one could better analyze the image.
> I don’t know what framing and sorting means. To me, it needs explanation or hyperlinks to definitions.
The results also showed that images had maximum non-terpene and pre-terpene pathways.
> Maximum in what way? Why is that significant?

Future Aspects:
> Rather than “Future Aspects”, perhaps “Next Steps”? What title would best describe all the bullet-points you end up including here?

Rules need modification.
Another round of annotations with specific tabel should be circulated.

**Comment by Vaish2611** [comment]
## **Project 2: To create C directory in the pyami.image consisting of 30 images of Terpene** 
In this proect we are creating a directory which we could simply run over by pyamiimage and extract named terpenes from the images. 

**Role:**
Created a list of 30 images which consist of list of links taking to the images that are definetely of Terpene Pathway. 
<img width="227" alt="image" src="https://user-images.githubusercontent.com/78601110/161524471-c14d4e9d-2a7c-4f7a-a1f4-937fe317fcb2.png">

## **Diretory FrameWork**

C:\Users\pyamiimage\test\alex_pico\image_30\PMC6707142\figure\Fig1\caption

## **Project 3: Manual Curation of pyamimage**

Have been working with the different terpenes images and am extracting words from it test out the pyamimage tool have created a file to spot out the difficulties or problems faced. The [tabel](https://docs.google.com/spreadsheets/d/166mgWUWQ2JWh8tgrN-LANw0n5qlBwOfFKSUOtsMBfeA/edit?usp=sharing) consists of the work.

**Reply by EmanuelFaria** [reply]
Project 2: To create C directory in the pyami.image consisting of 30 images of Terpene
In this proect we are creating a directory which we could simply run over by pyamiimage and extract named terpenes from the images.
Role: Created a list of 30 images which consist of list of links taking to the images that are definetely of Terpene Pathway. 
Diretory FrameWork
C:\Users\pyamiimage\test\alex_pico\image_30\PMC6707142\figure\Fig1\caption
Project 3: Manual Curation of pyamimage
Have been working with the different terpenes images and am extracting words from it test out the pyamimage tool have created a file to spot out the difficulties or problems faced. The tabel consists of the work.

**Comment by EmanuelFaria** [comment]
Project 2: Creating Dictionaries
A dictionary is a collection of words with their meanings and some additional information.
> Would it be more precise to say something along the lines of… 

>DEFINITION:
“In the context of this project’s requirements, dictionaries are collections of words that are all related to a specific topic. For example, “eoPlants.xml” is a dictionary including plant names IBotanical and common), wikidataID…. etc.”?
>Suggest the next line provides some context for their significance.
eg. USE: 
“Our dictionaries will be used by… to… so that…”

>It might also be useful to put in sub-sub-headings like.
 PROJECT FOCUS or SELECTED TOPIC?:
Terpenes are… 
(Perhaps a VERY short description on why they are important (synthesize compounds that can’t be made synthetically?), and how there are still many more to discover and add to the terpine pathway wiki?)
And if you think appropriate, consider swapping Project 2 for Project 1…to start with the greater context of the project, within which the sub-context of the project’s focus topic? Then here you can say “terpenes and their pathways are discussed in more detail below in Project 2” or something like that — if that makes sense to you.

> CREATION METHOD(S):
1) Manual Curation. 
(If you want to talk about that at all, a sentence or two would suffice, but it’s advantages and limitations would be interesting, to me anyway).

> 2)MACHINE CREATION/CURATION:?
2a) text analysis
2b) image analysis
ETC..

The words are extracted from images using 'optical character recognition' tools Tesseract and EasyOCR. Then, the words are looked up in the WikiData for their element code, description, and relatable element codes. This information is stored in dictionaries.
We are creating dictionaries with terpene and terpene synthase elements. Unlike other dictionaries, these will contain the frequency of the words as well as their synonyms.
Collecting words from image analysis tools
> Perhaps start with a description of the problem/challenge at this step: Relevant papers on this topic usually contain images that contain curtial information that are not included in the body text. 
This visual information includes… arrows/vectors, ….
We are developing our image analysis tool to solve that problem…
Here’s how…
> (More sub-sub-headings as useful to break the information into bite-sized chunks which also serve to make the document more “scannable” by the reader. (i.e. the Reader can get a good idea of where this paper is heading by scanning the headings before getting into the body text)

By integrating an optical character recognition tool(Tesseract) in Pyamiimage, we extract words from the terpene pathway images.
Let's take the figure given below as an example.
 The words obtained are:
\
Dictionary:
> Add labels and descriptive captions to the images, if helpful.


Fuzzywuzzy matching
Fuzzywuzzy is a string matching library in Python. As the words were not completely obtained in the output, this was used to match the names of various elements.
> Would be useful to hyperlink Python and any other jargon in the paper to layman-readable definitions.
> RE: “As the words were not completely obtained in the output, this was used to match the names of various elements.”
Does this mean that if the OCR comes up with garbage (like swapping 5 for S or 0 for O) that string matching would “learn/remember/correct” that?

> Here’s key format used in writing copy for marketing materials (such as sales letters) that might be useful here:
1. Tell me what you’re going to tell me.
2. Tell me
3. Tell me what you told me.

> I’m suggesting you consider: For each major paper section (topic change within the paper), provide n short intro that:
First, describes what you’re about to show them. Then show them with examples. And finally, restate the whole thing before moving to the next topic change… “As you can see thanks to (string matching), x problem as shown in y examples resulted in z useable output.”


Fuzzywuzzy test
> RE: fuzzywuzzy example image. Does 80 also mean there is an 80% chance that the fuzzy wuzzy result is correct? I have no idea. Just checking.

Cleaning data manually
> Regarding Sub-title “Cleaning data manually”… are these sub-heads actually “steps that are completed in order, or just components being highlighted? If steps, I suggest you name them as such: Step 1: X, Step 2: y…

The words acquired from the tool had a few errors, so the data was manually cleaned. (https://github.com/petermr/pyamiimage/blob/main/cleaned_phrases/Phrases.csv)
Code for Dictionaries
> RE: “Code for Dictionaries”… would something along the lines of “Coding for Dictionary Creation” be more precise? If your document can be understood solely by reading the headings and su-headings, you’re on the right track to naming them.

To create dictionaries, code in Python was formulated.
> is code “formulated” or “developed” or something else. Peter would know for sure.
 Dictionary code This code also calculates the frequency of occurrence of the words in the list.
> RE “This code also calculates the frequency of occurrence of the words in the list”

Suggest: “The code in the example below (Fig 2.0) does x number of things: 1, 2, 3.”

For example:
